packages/pn-validator/src/rules/ForEachElement.ts

```diff
@@ -19,38 +26,75 @@ export class ForEachElement<TModel, TValue> extends Rule<TModel, TValue> {
   }
 
-  private pushRule = (rule: Rule<TModel,  TElemValue<TValue>>) => {
+  private pushRule = (rule: Rule<TModel, TElemValue<TValue>>) => {
     this.rules.push({ isAsync: false, rule });
   };
 
-  private getTypeRules = (): TypeRules<TModel,  TElemValue<TValue>> =>
+  private isString = () => {
+    this.pushRule(new IsString());
+    return new StringRuleValidator<TModel, TElemValue<TValue>>(this.pushRule);
+  };
+
+  private isNumber = () => {
+    this.pushRule(new IsNumber());
+    return new NumberRuleValidator<TModel, TElemValue<TValue>>(this.pushRule);
+  };
+
+  private isDate = () => {
+    this.pushRule(new IsDate());
+    return new DateRuleValidator<TModel, TElemValue<TValue>>(this.pushRule);
+  };
+
+  private isBoolean = () => {
+    this.pushRule(new IsBoolean());
+    return new BooleanRuleValidator<TModel, TElemValue<TValue>>(this.pushRule);
+  }
+
+  private isObject = () => {
+    this.pushRule(new IsObject());
+    return new ObjectRuleValidator<TModel, TElemValue<TValue>>(this.pushRule);
+  };
+
+  private isArray = () => {
+    this.pushRule(new IsArray());
+    return new ArrayRuleValidator<TModel, TElemValue<TValue>>(this.pushRule);
+  };
+
+  private getTypeRules = (): TypeRules<TModel, TElemValue<TValue>> =>
     ({
-      ...new StringRuleValidator<TModel,  TElemValue<TValue>>(this.pushRule),
-      ...new NumberRuleValidator<TModel,  TElemValue<TValue>>(this.pushRule),
-      ...new DateRuleValidator<TModel,  TElemValue<TValue>>(this.pushRule),
-      ...new ObjectRuleValidator<TModel,  TElemValue<TValue>>(this.pushRule),
-      ...new ArrayRuleValidator<TModel,  TElemValue<TValue>>(this.pushRule),
-    } as unknown as TypeRules<TModel,  TElemValue<TValue>>);
+      isString: this.isString,
+      isNumber: this.isNumber,
+      isDate: this.isDate,
+      isBoolean: this.isBoolean,
+      isObject: this.isObject,
+      isArray: this.isArray,
+    } as unknown as TypeRules<TModel, TElemValue<TValue>>);
 
   public valueValidator = (value: TValue, model: TModel) => {
     if (!isDefined(value)) {
       return null;
     }
-    if (isArray<unknown, Array<unknown>>(value)) {
+    if (isArray(value)) {
       const results: Array<ValidationResult<TElemValue<TValue>>> = [];
       // eval rules for each element
       for (const rule of this.rules) {
+        let index = 0;
         for (const el of value) {
           const result = rule.rule.validate(el as TElemValue<TValue>, model);
 
           // check if there are errors
+          let resultToAdd = null;
           if (hasError(result)) {
-            results.push(result);
+            resultToAdd = result;
+          }
+          if (results[index] === undefined) {
+            results.push(resultToAdd);
             continue;
           }
-          results.push(null);
+          results[index] = resultToAdd;
+          index++;
         }
-        if (results.length > 0) {
+        if (results.filter(r => r === null).length > results.length) {
```

Comment:
I cannot imagine a scenario for this condition to be true, in general I guess that 
```
someArray.filter(someCondition).length <= someArray.length
```
holds for any array and for any condition. But maybe there is some case I am not aware of.

Reply:
yes, you are right!